File: java/src/main/java/sa/amnn/kit/webhooks/WebhookHandler.java
package sa.amnn.kit.webhooks;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Consumer;
import sa.amnn.kit.Models;

/** Framework-agnostic (Spring, Quarkus, servlet): call {@link #handle} with the headers and RAW body bytes.
 *  Verifies the {@code X-Webhook-Signature} over {@code X-Webhook-Timestamp + "." + rawBody}, de-duplicates by
 *  the delivery timestamp, then dispatches. Keep the consumer fast; queue heavy work. */
public final class WebhookHandler {
    public record Event(String id, String type, JsonNode data, byte[] raw) {}
    public record Result(int status, Map<String, Object> body) {}
    private final String secret; private final Consumer<Event> onEvent; private final Set<String> seen = Collections.synchronizedSet(new HashSet<>());
    public WebhookHandler(String secret, Consumer<Event> onEvent) { this.secret = secret; this.onEvent = onEvent; }

    public Result handle(Map<String, String> headers, byte[] rawBody) {
        String sig = header(headers, WebhookSignature.SIGNATURE_HEADER);
        String timestamp = header(headers, WebhookSignature.TIMESTAMP_HEADER);
        if (!WebhookSignature.verify(secret, timestamp, rawBody, sig)) return new Result(401, Map.of("error", "invalid signature"));
        JsonNode data;
        try { data = Models.JSON.readTree(rawBody); if (data == null || !data.isObject()) throw new IllegalArgumentException(); } catch (Exception e) { return new Result(400, Map.of("error", "invalid json")); }
        String type = data.hasNonNull("event") ? data.get("event").asText() : data.path("type").asText(null);
        String id = data.hasNonNull("timestamp") ? data.get("timestamp").asText() : String.valueOf(timestamp);
        if (!id.isEmpty() && !seen.add(id)) return new Result(200, Map.of("ok", true, "duplicate", true));
        onEvent.accept(new Event(id, type, data, rawBody));
        return new Result(200, Map.of("ok", true));
    }

    private static String header(Map<String, String> headers, String name) {
        return headers.entrySet().stream().filter(e -> e.getKey().equalsIgnoreCase(name)).map(Map.Entry::getValue).findFirst().orElse(null);
    }
}

File: kotlin/amen-client/src/main/kotlin/sa/amnn/kit/webhooks/WebhookHandler.kt
package sa.amnn.kit.webhooks

import kotlinx.serialization.json.*
import sa.amnn.kit.json
import java.util.Collections

data class WebhookEvent(val id: String, val type: String?, val data: JsonObject, val raw: ByteArray)
data class WebhookResult(val status: Int, val body: Map<String, Any>)

/** Framework-agnostic (Ktor, Spring, http4k): call [handle] with the headers and RAW body bytes.
 *  Verifies the real Amen signature scheme first (timestamp + "." + rawBody), de-duplicates by the
 *  top-level event timestamp (no event id exists in the body), then dispatches. Keep [onEvent] fast;
 *  queue heavy work. */
class WebhookHandler(private val secret: String, private val onEvent: suspend (WebhookEvent) -> Unit) {
    private val seen: MutableSet<String> = Collections.synchronizedSet(HashSet())

    private fun header(headers: Map<String, String>, name: String): String? =
        headers.entries.firstOrNull { it.key.equals(name, ignoreCase = true) }?.value

    suspend fun handle(headers: Map<String, String>, rawBody: ByteArray): WebhookResult {
        val sig = header(headers, WebhookSignature.SIGNATURE_HEADER)
        val timestamp = header(headers, WebhookSignature.TIMESTAMP_HEADER)
        if (!WebhookSignature.verify(secret, timestamp, rawBody, sig)) return WebhookResult(401, mapOf("error" to "invalid signature"))
        val data = runCatching { json.parseToJsonElement(rawBody.decodeToString()).jsonObject }.getOrNull() ?: return WebhookResult(400, mapOf("error" to "invalid json"))
        val type = (data["event"]?.jsonPrimitive?.contentOrNull) ?: header(headers, WebhookSignature.EVENT_HEADER)
        val id = data["timestamp"]?.jsonPrimitive?.contentOrNull ?: timestamp ?: ""
        if (id.isNotEmpty() && !seen.add(id)) return WebhookResult(200, mapOf("ok" to true, "duplicate" to true))
        onEvent(WebhookEvent(id, type, data, rawBody))
        return WebhookResult(200, mapOf("ok" to true))
    }
}
